### PR TITLE
GH-36479: [C++][FlightRPC] Use gRPC version detected by find_package()

### LINKED ISSUE #36479: [C++][FlightRPC] Use gRPC version macros if possible
### Describe the enhancement requested

Currently, we have a suite of compilation-based configuration tests to find out the approximate gRPC version, for feature selection.

Starting with 1.51.0 however, the gRPC C++ API exposes version macros:
https://github.com/grpc/grpc/pull/31033

We could start using those and fallback on the older method if the macros are not available.

### Component(s)

C++, FlightRPC

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

We don't need to use `try_compile()` by using gRPC version detected by `find_package()`.

### What changes are included in this PR?

Use  gRPC version detected by `find_package()`.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* Closes: #36479

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Review by lidavidm — APPROVED** [review]
This is a great cleanup!

File: cpp/cmake_modules/FindgRPCAlt.cmake [hunk 1]
```diff
@@ -45,6 +47,7 @@ if(GRPCPP_PC_FOUND)
                    HINTS ${GRPCPP_PC_STATIC_LIBRARY_DIRS})
       list(APPEND GRPCPP_LINK_LIBRARIES "${GRPCPP_LIBRARY_${GRPCPP_LIBRARY_NAME}}")
     endforeach()
+    list(APPEND GRPCPP_LINK_LIBRARIES Threads::Threads)
```
Comment (assignUser) [inline cpp/cmake_modules/FindgRPCAlt.cmake:50]: This is duplicated from L39, should we move it under the if as it is appended to the list in all cases?

**Review by assignUser — COMMENTED** [review]
Nice clean up, one nit :)

File: cpp/cmake_modules/FindgRPCAlt.cmake (hunk 1 above)
Comment (assignUser) [reply cpp/cmake_modules/FindgRPCAlt.cmake:50]: (I apparently had this open while it was merged ^^ so we can do it in another related PR)

Comment (kou) [reply cpp/cmake_modules/FindgRPCAlt.cmake:50]: Oh, sorry. Could you work on this?

File: cpp/src/arrow/flight/CMakeLists.txt [hunk 2]
```diff
@@ -102,84 +102,21 @@ set(CMAKE_CXX_FLAGS_BACKUP "${CMAKE_CXX_FLAGS}")
 string(REPLACE "/WX" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
 string(REPLACE "-Werror " " " CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
 
-# Probe the version of gRPC being used to see if it supports disabling server
```
Comment (pitrou) [inline cpp/src/arrow/flight/CMakeLists.txt:105]: @kou Is there a reason we kept the KLUDGE above when removing the `try_compile` calls?

Comment (kou) [reply cpp/src/arrow/flight/CMakeLists.txt:105]: No. I should have removed the KLUDGE too in this PR.

**Pull request merged**